### ISSUE TITLE
`after:` and `before:` date search terms are underspecified

### ISSUE THREAD

**Issue by nikclayton** [mastodon/mastodon#31142]
### Steps to reproduce the problem

Filing this mostly to get the information down somewhere so it's searchable and can inform improvements in the next iteration of the search API.

From inspection, the `after:` and `before:` search operators:

- They only take yyyy-mm-dd formatted dates, there's no mechanism to specify hour/minutes or a specific timezone.
- They imply a time, but the time is different depending on the operator. The `after:` operator behaves as though the time was 23:59:59.999... on the date, the `before:` operator behaves as though the time was 00:00:00 on the the date.

This means the ranges they specify are exclusive rather than inclusive. I.e. `after:2024-07-15` will start from midnight on 2024-07-16, `before:2024-07-15` will end at midnight on 2024-07-14.

A query of `after:2024-07-01 before:2024-07-02` will return nothing. It would make more sense if the missing time component was set to 00:00:00 in both cases, and then this would return statuses from July 1st. I'm aware of the `during:` operator, but that just means clients have to special-case this (or expand any date ranges entered by the user by one day on either end).

- The input dates are not treated according to the user's timezone, which leads to inconsistencies in the presentation.

On that last point, I'm in Zurich, so UTC+2.

If I go to the web UI and enter the following search query freebsd after:2024-06-30 before:2024-07-02 (the intent being to only show me posts made on 1st July) the first result is https://mastodon.social/@kurtseifried@infosec.exchange/112713824626524406.

This was posted at 2024-07-01T23:27:53.000Z according to the API. But in the UI that's rendered as "Jul 2". So from the user's perspective this status should not included in the search results.

### Expected behaviour

N/A

### Actual behaviour

N/A

### Technical details

If this is happening on your own Mastodon server, please fill out those:
- Ruby version: (from `ruby --version`, eg. v3.1.2)
- Node.js version: (from `node --version`, eg. v18.16.0)

**Comment by notchairmk** [comment]
Just leaving some notes:

> - They only take yyyy-mm-dd formatted dates, there's no mechanism to specify hour/minutes or a specific timezone.

Supporting datetimes on top of dates would be very nice to have from a search perspective

> - They imply a time, but the time is different depending on the operator. The after: operator behaves as though the time was 23:59:59.999... on the date, the before: operator behaves as though the time was 00:00:00 on the the date.

Time zone _should be_ respected if you have it set on your account: https://github.com/mastodon/mastodon/blob/f6fb9b75dbc10cee8beee4bcb84dd08e821baf1b/app/lib/search_query_transformer.rb#L168-L175

But could also be something that's inherited from the user's browser locale.

---

In my mind, the best way of starting to go about building out additional date search functionality would be to start by adding validation (started on in #31526). And then adding additional functionality on top of that. Would be happy to take a look if this is something that the project wants to do.

**Comment by b123400** [comment]
It looks like time specific search can be done via `min_id` and `max_id`, as Mastodon uses snowflake IDs.
https://shkspr.mobi/blog/2022/11/building-an-on-this-day-service-for-mastodon/

You can try this in your browser console
```JavaScript
function datetimeToSnowflake(date) {
  return BigInt(date.getTime()) << 16n;
}

function snowflakeToDatetime(id) {
  return new Date(Number(BigInt(id) >> 16n));
}
```

**Comment by nikclayton** [comment]
> It looks like time specific search can be done via `min_id` and `max_id`, as Mastodon uses snowflake IDs.

While this is technically correct, it's a bad idea:

> Always treat IDs as opaque strings.
> -- https://docs.joinmastodon.org/api/guidelines/#id

**Comment by b123400** [comment]
You are right. I tried to use `min_id` and `max_id` as if they are effectively `before` and `after`. I then realised the timestamp in the snowflake ID is often different from the `created_at` of a status. The Snowflake ID is generated when the server first sees a status, which might be someone reposting a status created few days ago on another server.
So yes, it would be great if `before` and `after` support precise timestamp.
It would be even better if it's also supported in the `/api/v1/timelines/public` endpoint.

**Comment by trwnh** [comment]
The other issue with trying to parse dates out of Snowflake IDs is that the Mastodon API `id` is not guaranteed to be a Snowflake ID. The official Mastodon software uses Snowflake ID currently, but it could change to any other ID scheme later, and unofficial implementations of the Mastodon API already switched to non-Snowflake schemes years ago.
